### ISSUE TITLE
[BUG] Rancher 2.7.5 Active Directory authentication fails completely in Rancher 2.7.5 due to wrong objectGUID escaping

### ISSUE THREAD

**Issue by FFock** [rancher/rancher#41985]
**Rancher Server Setup**
- Rancher version: 2.7.5
- Installation option (Helm Chart): K8S 1.24.x RKE1
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version: 1.24.x
- Cluster Type (Local/Downstream): Local

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - All users are affected!

**Describe the bug**
After upgrading to Rancher 2.7.5 (from 2.7.4) LDAP based login using Active Directory integration fails for all users.

**To Reproduce**
Upgrade to 2.7.5 and then login with a fresh browser.

**Result**
Login failure. In Rachner log (GUID data removed):
`LDAP Result Code 201 Filter Compile Error: ldap: invalid characters for escape in filter: encoding/hex: invalid byte: U+004E N`

**Expected Result**
Login success

**Additional context**
The following change has introduced to problem:
https://github.com/rancher/rancher/commit/cdd90ea6c687d8c83cfd0289e51be32da5c21d14

I think the problem source is that `guidString := html.EscapeString(fmt.Sprintf("%x", entry.GetRawAttributeValue(common.AttributeObjectGUID)))` is used to build an URL from the `objectGUID `value. This might result in a string that is not fully hex encoded. All our `objectGUIDs` contain the "N" byte with hex value '4E'. The HTML escaping might not escape this byte and then when it is used later in LDAP the filter string will contain "...\N..." which is not a proper escaping and leads to the above noted error message. This is still an assumption, because I have not understood the relevant code fully yet, but I am pretty sure, that it points at least in the right direction.

**Comment by crobby** [comment]
@FFock Do you have an example GUID from a test system or any log messages?  Just an example of the format is necessary, I don't need a real one.

**Comment by daleckystepan** [comment]
Hi,
I just updated to Rancher v2.7.5 from v2.7.2 and this change completely break auth against AD. I don't expect breaking change in patch update. I am using: https://registry.terraform.io/providers/rancher/rancher2/latest/docs/resources/auth_config_activedirectory
which uses activedirectory_group://<DN> or activedirectory_user://<DN> and any other resources use the same approach. Any idea?

It is not manageable to use GUIDs here.

Thank you.

**Comment by FFock** [comment]
> @FFock Do you have an example GUID from a test system? Just an example of the problematic GUID is necessary, I don't need a real one. My test system may be creating GUIDs in a different way, so I want to be sure that the fix will cover all possible GUIDs that we will encounter.

@crobby The format seems to be very simple: 
16-n bytes (the length varies between users), where the sequence "ef bf bd" (hex) occurs very often. Inbetween there is binary data where some of the bytes represent ASCII characters.

**Comment by FFock** [comment]
This is bug is more fatal than I thought initially, because most users cannot login even after having rolled back to Rancher 2.7.4. It seems that a migration of the user identifier from `DN `to `objectGUID `had already been running in the background and now all migrated users cannot login anymore. 

**We really need very urgently a fix for this.**

**Comment by FFock** [comment]
Even after having restored the Rancher local cluster to the state before the 2.7.5 upgrade does not fix the login issue. We now have a lot of users with GUID user identifier, which cannot be deleted:
`[ERROR] Unknown error: DN ended with incomplete type, value pair`
On the other hand, readding the users with their DN user identifier, **does not work either**:
```
[INFO] Creating user for principal activedirectory_user://CN=Name\, Forename (123456),OU=99,OU=orgunit,DC=somedc,DC=com
2023/07/04 10:04:08 [ERROR] error syncing 'c-s484l/crtb-4n3cp': handler mgmt-auth-crtb-controller: users.management.cattle.io "u-4g51sghhql" already exists, handler auth-prov-v2-crtb: roletemplatebinding doesn't have any subject fields set: crtb-4n3cp, requeuing
```

**Comment by 0Styless** [comment]
We also encounter this bug after upgrading from 2.7.4 to 2.7.5 We have a lot of users that have broken permissions or can't log into Rancher at all.

I agree that this is a major bug! We also opened a critical issue in the SUSE customer center.

**Comment by 0Styless** [comment]
It is also no longer possible to add users to projects, here is the error message in the frontend:

![image](https://github.com/rancher/rancher/assets/30753608/839bf9d7-dd02-4a0f-b063-f7c129ec6542)

**Comment by betweenclouds** [comment]
We have the same here after going from 2.7.3 to 2.7.5.
But we are still able to add AD Groups and users in AD Groups are still able to login.

Opened a issue with SUSE too.

**Comment by JacieChao** [comment]
https://github.com/rancher/rancher/issues/37994#issuecomment-1610906320

**Comment by FFock** [comment]
> Even after having restored the Rancher local cluster to the state before the 2.7.5 upgrade does not fix the login issue. We now have a lot of users with GUID user identifier, which cannot be deleted:
> `[ERROR] Unknown error: DN ended with incomplete type, value pair`

There is a workaround for the above problem, users with GUID identifier have to be deleted on "Users & Authentication" and then the corresponding entries on down-stream cluster are deleted automatically too. After that, you can re-add those users on downstream cluster level again. Then those are again working with DN identifier on Rancher 2.7.4.

**Comment by 0Styless** [comment]
> > Even after having restored the Rancher local cluster to the state before the 2.7.5 upgrade does not fix the login issue. We now have a lot of users with GUID user identifier, which cannot be deleted:
> > `[ERROR] Unknown error: DN ended with incomplete type, value pair`
> 
> There is a workaround for the above problem: Users with GUID identifier have to be deleted on "Users & Authentication" and then the corresponding entries on down-stream cluster are deleted automatically too. After that, you can re-add those users on downstream cluster level again. Then those are again working with DN identifier on Rancher 2.7.4.

I can confirm that this is a workaround for users who can't log in - but not a satisfying one. You will have to re-configure single permissions to projects on all downstream clusters.

**Comment by FFock** [comment]
We now are facing another related problem: Users which had been affected by this issue during the 2.7.4->2.7.5 upgrade attempt cannot use scoped API keys (tokens) anymore after the rollback of the local Rancher HA 2.7.5 cluster to 2.7.4 through server restore.

**Comment by AndrewHoffmanQA** [comment]
QA notes:

QA scenarios-
- user upgrades from 2.7.4 to 2.7.5 with nested OUs and users in it- confirm nested OU compatibility 
- user upgrades to 2.7.5 and then creates a nested OU and create users into the nested OU. log in as user
- user upgrades to 2.7.5 and then edits a nested OU and moves users around OU. log in as user
- user upgrades to 2.7.5 and then deleted a nested OU and move the users to the base OU. log in as user

**Comment by daleckystepan** [comment]
Please take also API changes into consideration. Currently it breaks all scripts and/or terraform modules because it requires GUID instead of DN. GUIDS in API or terraform are very hard to maintain.

Thank you

**Comment by geoff-carr-bzy** [comment]
Very frustrating to have hit this bug having been forced to upgrade to `2.7.5` from `2.7.3` due to the OOM bug, in additional to having to hold off upgrading to K8S `1.24` due to the high CPU bug! Seems to be a very buggy trend here!

**Comment by papanito** [comment]
We are also suffering this issue

**Comment by geoff-carr-bzy** [comment]
> We are also suffering this issue and can see for each user logging in a new account is created.

Deleting the older account seems to restore the ability to login to UI for the user, but then all stored KUBECONFIG will need to be updated as the tokens will then be invalid.

**Comment by papanito** [comment]
@geoff-carr-bzy yes exactly and I don't really want to ask all users to update their KUBECONFIG due to a bug. 
On the other hand, if you delete the newer one and the users try again to login in the UI, this also works.

**Comment by mhoffmann01** [comment]
Same problem here. Since the update to 2.7.5, I couldn't add existing and new users to projects. I also got the error message 'No identities can be retrieved'. I were able to add groups to the 'who is authorized to login' part under 'Auth-Provider', but not a single user, just groups.

In the logs of the rancher pods, I saw that it searched for something like this: (&(objectClass=person)(guid=`\`..`\`..`\`..`\`)) where the `\` parts seem to be escaped characters. A little confusing is, that the actual guid is something like `aabbccdd-aabbcc-aabbccdd`, but the escaped and searched one was like `ddccbbaa-ccbbaa-ddccbbaa` ... is this the right thing to do?

I switched to the `OpenLDAP` auth provider. Regarding the [source code](https://github.com/rancher/rancher/blob/release/v2.7/pkg/auth/providers/ldap/ldap_client.go), it does things differently for `getPrincipal` and `RefetchGroupPrincipals` by do not setting the ldap search level to `SingleLevel`. Is this maybe a thing?
After the switch, everything worked as before, but I had to set the permission per user per project again... manually. 

I hope it helps somehow.

**Comment by frjaraur** [comment]
Same here, we are stuck in 'No identities can be retrieved'. Any good workaround, besides moving authentication to AD groups (currently they are working)?

**Comment by mhoffmann01** [comment]
As I wrote, I changed the auth provider to OpenLDAP and can (again) authorize through users and groups, as well as assign single users to groups etc. But I can think of scenarios where this provider change isn't suitable, since all existing permissions are gone afterwards.

**Comment by frjaraur** [comment]
As far as I understand, changing the auth provider will need to recreate all the current projects users and groups management; which is not possible in a production environment.

**Comment by mhoffmann01** [comment]
Exactly. As far as I understand from the last commit of the [source code](https://github.com/rancher/rancher/commit/cdd90ea6c687d8c83cfd0289e51be32da5c21d14), they changed the ldap search scope. I don't know why, but maybe changing this back, could solve the problem without changing the auth provider.

**Comment by nobbs** [comment]
We're also facing the same issue with our ActiveDirectory auth backend setup. To summarize, as I haven't seen all of the issues reported in one comment:

- Rancher tried to do the conversion of users with the old DN-based principalIds to the new GUID one. For some users it worked, so that these now only have the new one. For some it apparently failed, ending up with two users for the same person - the old DN-based one and the new GUID-based on. 
- Project and Cluster role template bindings were not properly updated for the users where the migration failed, some are still pointing to old DN-based users.
- _My guess_: when a user tries to log in, the auth backend is using the new GUID-based users regardless whether conversion failed or not. As not all role template bindings were updated, these users lose their permissions as the bindings are still referencing the old DN-based users. 
- _Why exactly it's failing I'm not sure._ We're seeing the following logs repeating all the time for different IDs - some of those log entries are probably also caused by users being removed from AD at some point but the role binding still being there.

```shell
rancher-55d968698d-lm487 rancher 2023/07/12 11:56:52 [ERROR] Error refreshing token principals, skipping: Cannot locate user information for (&(objectClass=person)(objectGUID=\c0\a5\6a\7d\a2\00\2e\4d\a2\07\e6\15\59\37\99\c5))
rancher-55d968698d-lm487 rancher 2023/07/12 11:56:52 [ERROR] error syncing 'user-2jtvq': handler mgmt-auth-userattributes-controller: No identities can be retrieved, requeuing
rancher-55d968698d-lm487 rancher 2023/07/12 11:56:53 [ERROR] Error refreshing token principals, skipping: Cannot locate user information for (&(objectClass=person)(objectGUID=\4f\60\0f\16\59\59\73\46\a7\1e\0a\8c\a5\95\b8\8a))
rancher-55d968698d-lm487 rancher 2023/07/12 11:56:54 [ERROR] error syncing 'u-ii5rhjptp4': handler mgmt-auth-userattributes-controller: No identities can be retrieved, requeuing
rancher-55d968698d-lm487 rancher 2023/07/12 11:56:59 [ERROR] Error refreshing token principals, skipping: Cannot locate user information for (&(objectClass=person)(objectGUID=\70\fa\14\2e\3b\03\82\4f\b0\5f\14\43\ea\d0\2e\c5))
rancher-55d968698d-lm487 rancher 2023/07/12 11:56:59 [ERROR] error syncing 'u-3pf7cvyg6r': handler mgmt-auth-userattributes-controller: No identities can be retrieved, requeuing
rancher-55d968698d-lm487 rancher 2023/07/12 11:57:08 [ERROR] Error refreshing token principals, skipping: Cannot locate user information for (&(objectClass=person)(objectGUID=\9e\d8\ac\74\d3\84\0f\41\9d\d7\fe\be\2b\0c\a5\db))
rancher-55d968698d-lm487 rancher 2023/07/12 11:57:09 [ERROR] error syncing 'u-6fcsvou6am': handler mgmt-auth-userattributes-controller: No identities can be retrieved, requeuing
rancher-55d968698d-hflzl rancher 2023/07/12 11:57:11 [ERROR] Unknown error: No identities can be retrieved
rancher-55d968698d-lm487 rancher 2023/07/12 11:57:11 [ERROR] Error refreshing token principals, skipping: Cannot locate user information for (&(objectClass=person)(objectGUID=\01\1c\d2\16\77\b2\e4\48\94\37\fd\8f\25\54\20\79))
rancher-55d968698d-lm487 rancher 2023/07/12 11:57:12 [ERROR] error syncing 'u-kbcfbecaef': handler mgmt-auth-userattributes-controller: No identities can be retrieved, requeuing
rancher-55d968698d-lm487 rancher 2023/07/12 11:57:12 [ERROR] Error refreshing token principals, skipping: Cannot locate user information for (&(objectClass=person)(objectGUID=\b1\f0\0c\ef\2b\2f\62\4c\8a\2c\0c\cb\4c\55\ef\e4))
rancher-55d968698d-lm487 rancher 2023/07/12 11:57:12 [ERROR] error syncing 'u-kioxa7doxo': handler mgmt-auth-userattributes-controller: No identities can be retrieved, requeuing
rancher-55d968698d-lm487 rancher 2023/07/12 11:57:22 [ERROR] Error refreshing token principals, skipping: Cannot locate user information for (&(objectClass=person)(objectGUID=\56\11\86\ac\9b\9c\8b\45\ae\09\3e\c7\a2\02\0e\15))
rancher-55d968698d-lm487 rancher 2023/07/12 11:57:23 [ERROR] error syncing 'u-kutrdremqy': handler mgmt-auth-userattributes-controller: No identities can be retrieved, requeuing
rancher-55d968698d-lm487 rancher 2023/07/12 11:57:25 [ERROR] Unknown error: No identities can be retrieved
...
rancher-55d968698d-lm487 rancher 2023/07/12 11:57:46 [ERROR] error syncing 'u-7ff5ktddfv': handler mgmt-auth-userattributes-controller: No identities can be retrieved, requeuing
rancher-55d968698d-lm487 rancher 2023/07/12 11:58:30 [ERROR] Error refreshing token principals, skipping: Cannot locate user information for (&(objectClass=person)(objectGUID=\c9\93\8b\70\00\1f\41\4e\88\b9\40\7f\ae\c4\a0\54))
rancher-55d968698d-lm487 rancher 2023/07/12 11:58:30 [ERROR] error syncing 'u-ln4b4zpomm': handler mgmt-auth-userattributes-controller: No identities can be retrieved, requeuing
rancher-55d968698d-lm487 rancher 2023/07/12 11:58:32 [ERROR] Error refreshing token principals, skipping: LDAP Result Code 201 "Filter Compile Error": ldap: invalid characters for escape in filter: encoding/hex: invalid byte: U+004E 'N'
rancher-55d968698d-lm487 rancher 2023/07/12 11:58:32 [ERROR] error syncing 'u-d7v2tqctgm': handler mgmt-auth-userattributes-controller: NotFound 404: <redacted> not found, requeuing
rancher-55d968698d-lm487 rancher 2023/07/12 11:58:34 [ERROR] Error refreshing token principals, skipping: LDAP Result Code 201 "Filter Compile Error": ldap: invalid characters for escape in filter: encoding/hex: invalid byte: U+004E 'N'
rancher-55d968698d-lm487 rancher 2023/07/12 11:58:34 [ERROR] error syncing 'u-3uatfpifxg': handler mgmt-auth-userattributes-controller: NotFound 404: <redacted> not found, requeuing
rancher-55d968698d-lm487 rancher 2023/07/12 11:58:39 [ERROR] Error refreshing token principals, skipping: LDAP Result Code 201 "Filter Compile Error": ldap: invalid characters for escape in filter: encoding/hex: invalid byte: U+004E 'N'
rancher-55d968698d-lm487 rancher 2023/07/12 11:58:40 [ERROR] error syncing 'u-ncdwezl7ux': handler mgmt-auth-userattributes-controller: NotFound 404: <redacted> not found, requeuing
rancher-55d968698d-lm487 rancher 2023/07/12 11:58:41 [ERROR] Error refreshing token principals, skipping: LDAP Result Code 201 "Filter Compile Error": ldap: invalid characters for escape in filter: encoding/hex: invalid byte: U+004E 'N'
```

**Comment by travisghansen** [comment]
Does downgrading try to convert back to the old principal IDs? My installs seem to be completely messed up at this point even after downgrading etc :(

**Comment by Turb0Fly** [comment]
We are also affected.  Fortunately, we upgraded out test environment first.  We did postpone our production upgrade due to Kubernetes 1.24 high CPU bug with RKE and then the cattle cluster agent memory leak in 2.7.2+.  Was hoping on 2.7.5, but this is a real show stopper.

**Comment by travisghansen** [comment]
For my part none of them appear to have non-ASCII chars in the DN.

**Comment by philipp1992** [comment]
how can it be that this issue exists for 2 weeks and no hotfix has been released? my client is also affected by this.

**Comment by FFock** [comment]
> Is this affecting all users or only ones with non-ASCII characters in their DN?

No, as I wrote above in this ticket, it seems that the GUID migration/usage is failing if the GUID of a user contains at least one byte that represents a character that does not need to be escaped in an URL (i.e., some letter like 'N' or 'e').

**Comment by philipp1992** [comment]
@crobby  please revert the change that caused this issue and release a fix asap

**Comment by FFock** [comment]
> @crobby please revert the change that caused this issue and release a fix asap

I think at least for all users, that have already tried to install 2.7.5 somewhere with several downstream clusters active, reverting back to 2.7.4 will not solve the issue created by the upgrade. I did that already using a full HA Rancher 2.7.4 pre-upgrade backup, but the issue now is in the API key handling/tokens of the downstream cluster users (**scoped keys are no longer working for users affected by failed GUID migration - even after deletion and recreation of those users!**). 

So far, we have not found and are not aware of any workaround. So we need a forward fix that restores full AD functionality with 2.7.6+.

**Comment by 0Styless** [comment]
> > @crobby please revert the change that caused this issue and release a fix asap
> 
> I think at least for all users, that have already tried to install 2.7.5 somewhere with several downstream clusters active, reverting back to 2.7.4 will not solve the issue created by the upgrade. I did that already using a full HA Rancher 2.7.4 pre-upgrade backup, but the issue now is in the API key handling/tokens of the downstream cluster users (**scoped keys are no longer working for users affected by failed GUID migration - even after deletion and recreation of those users!**).
> 
> So far, we have not found and are not aware of any workaround. So we need a forward fix that restores full AD functionality with 2.7.6+.

I agree, just reverting these settings won't help any customer that has already upgraded Rancher to 2.7.5. The fix must contain a logic that reverts the already migrated user accounts to their original state. Or at least there should be a release without this change to benefit from the fixes with the cluster agent and cilium bugs there were included in the 2.7.5.

**Comment by Turb0Fly** [comment]
Is this being actively investigated ? We were eagerly awaiting this release to fix the high memory consumption bug on Rancher pod and cattle-cluster-agents, as well as introducting k8s 1.25 support to our teams.  We won't deploy 2.7.5 on our production instance because of this issue.  Still running 2.6.6 as it is the most stable release we experienced so far.

**Comment by philipp1992** [comment]
@crobby  could you please make a statement

**Comment by sowmyav27** [comment]
moving it to Review after an offline chat with @crobby. Waiting for the PR to be merged ^

**Comment by AndrewHoffmanQA** [comment]
QA Test plan WIP:


the bug-
in 2.7.5 we made a change to the principle field on AD users, switching them from distinguished name to GUID. for a couple reasons this process can fail. if the user tries to login before their principal has been migrated, they will lose their permissions.

reproduction steps- 
1. stand up a 2.7.4 rancher
2. configure AD
3. log in with multiple nested OU users
4. give the users permissions to projects
5. delete one of the users out of the AD
6. upgrade the server to 2.7.5
7. log in as one of the immigrated users 
  a. to check which users have been migrated, open kubectl
`kubectl get users -o yaml | grep activedirectory`
 b. located a user with a principle ID that hasn't been migrated
8. log in with the user that failed to migrate

bug- user does not have proper permissions and user is duplicated





bug fixes that should be verified-
1. test reproduction steps above, validate that the migration works as expected and users being deleted out of ADs do not result in a migration failure. confirm user is not duplicated
2. test reproduction steps above with a high volume of users (500). -- confirm if we will add release note saying how long much time to give the migration before attempting login. confirm all users migrate their IDs as expected.
3. how to test the cleanup script tool. work with Chad once he has created it. 

targeted regression areas-

**Comment by AndrewHoffmanQA** [comment]
Bug-
in 2.7.5 we made a change to the principle field on AD users, switching them from distinguished name to GUID. for a couple reasons this process can fail. if the user tries to login before their principal has been migrated, they will lose their permissions.

reproduction steps-

- stand up a 2.7.4 rancher
- configure AD
- log in with multiple nested OU users
- give the users permissions to projects "p1"
- delete one of the users out of the AD
- upgrade the server to 2.7.5
- log in as one of the immigrated users
- a. to check which users have been migrated, open kubectl
- kubectl get users -o yaml | grep activedirectory
- confirm logged in as migrated user works as expected

- locate a user with project permissions and with a principle ID that hasn't been migrated
- log in with the user that failed to migrate
- confirm the user doesn't have project permissions and is duplicated.



Bug
- user does not have proper permissions and user is duplicated
- logs throw this error `Login failure. In Rancher log (GUID data removed):
LDAP Result Code 201 Filter Compile Error: ldap: invalid characters for escape in filter: encoding/hex: invalid byte: U+004E`

bug repro documentation
- per screenshot, you can see we stopped the migration causing only some principle IDs to be converted. further down the screenshot shows user10's display issues due to failed migration, as well as its lack of project permissions when logged in as duser10. additionally there is a pic of the error code in the client reported bug, as well as picture proof of duser10 duplicating upon attempting to log in. 

<img width="2317" alt="image" src="https://github.com/rancher/rancher/assets/89030218/73f882a0-9e9f-406f-b155-1988fb5f0638">

**Comment by prachidamle** [comment]
Here are the revert PRs:
PRs:
https://github.com/rancher/rancher/pull/42169
https://github.com/rancher/rancher/pull/42167
https://github.com/rancher/rancher/pull/42168

And the issue for reverting the changes: https://github.com/rancher/rancher/issues/42161

**Comment by travisghansen** [comment]
Does anyone have an interim solution for this? It seems like I can't go backward and while we're waiting for 2.7.6 there's nothing to go forward to.

**Comment by nobbs** [comment]
> Does anyone have an interim solution for this? It seems like I can't go backward and while we're waiting for 2.7.6 there's nothing to go forward to.

That won't help you at all, but just for completeness sake: we've switched auth to AzureAD as our enterprise also makes use of Azure and has synchronisation set up. AAD doesn't have this issue, but we obviously still had to rebuild all permissions. Also has the positive side effect of giving us the option to enforce MFA (from AAD side).

**Comment by travisghansen** [comment]
What won't help me?

**Comment by nobbs** [comment]
> What won't help me?

My comment how we solved this for us.

**Comment by papanito** [comment]
@AndrewHoffmanQA I installed 2.7.6-rc5 (form 2.7.5) and then run scenario 1. You say 
> his display all user's principle IDs, confirm they are GUID not DS. GUID ID example- 4d80209a63f84309be662a208d5613f0

In my case, the users have a GUID. Still I can successfully login with my user and have the required permissions.

I also delete my 2nd user account, login again and this also only creates one user (but with a DS, not a GUID)

**Comment by papanito** [comment]
We updated to Rancher 2.7.6 and the issue seems fixed. Users are there only once using a DS (instead the new GUID)

**Comment by Turb0Fly** [comment]
Here as well so far on our test environment.  Updated to 2.7.6 and all AD users can now log in. They are also seen in Rancher by DN instead of GUID.  They can also be added to RBAC without any problem.
